src/stream_processor/flb_sp.c

```diff
@@ -526,6 +523,361 @@ struct flb_sp *flb_sp_create(struct flb_config *config)
+        break;
+    }
+}
+
+static bool value_to_bool(struct flb_exp_val *val) {
+    bool result;
+
+    switch (val->type) {
+    case FLB_EXP_BOOL:
+        result = val->val.boolean;
+        break;
+    case FLB_EXP_INT:
+        result = val->val.i64 > 0;
+        break;
+    case FLB_EXP_FLOAT:
+        result = val->val.f64 > 0;
+        break;
+    case FLB_EXP_STRING:
+        result = true;
+        break;
+    }
+
+    return result;
+}
+
+
+static void logical_operation(struct flb_exp_val *left,
+                              struct flb_exp_val *right,
+                              struct flb_exp_val *result, int op)
+{
+    bool lval;
+    bool rval;
+
+    lval = left? value_to_bool(left) : false;
+    rval = right? value_to_bool(right) : false;
+
+    result->type = FLB_EXP_BOOL;
+
+    switch (op) {
+    case FLB_EXP_NOT:
+        result->val.boolean = !lval;
+        break;
+    case FLB_EXP_AND:
+        result->val.boolean = lval & rval;
+        break;
+    case FLB_EXP_OR:
+        result->val.boolean = lval | rval;
+        break;
+    }
+}
+
+static struct flb_exp_val *reduce_expression(struct flb_exp *expression,
+                                             msgpack_object *map)
+{
+    int operation;
+    struct flb_exp_val *ret, *left, *right;
+    struct flb_exp_val *result;
+
+    if (!expression) {
+        return NULL;
+    }
+
+    result = flb_malloc(sizeof(struct flb_exp_val));
+    if (!result) {
+       flb_errno();
+       return NULL;
+    }
+
+    switch (expression->type) {
+    case FLB_EXP_BOOL:
+        result->type = expression->type;
+        result->val.boolean = ((struct flb_exp_val *) expression)->val.boolean;
+        break;
+    case FLB_EXP_INT:
+        result->type = expression->type;
+        result->val.i64 = ((struct flb_exp_val *) expression)->val.i64;
+        break;
+    case FLB_EXP_FLOAT:
+        result->type = expression->type;
+        result->val.f64 = ((struct flb_exp_val *) expression)->val.f64;
+        break;
+    case FLB_EXP_STRING:
+        result->type = expression->type;
+        result->val.string = ((struct flb_exp_val *) expression)->val.string;
+        break;
+    case FLB_EXP_KEY:
+        ret = key_to_value(((struct flb_exp_key *) expression)->name, map);
+        if (!ret) {
+           flb_free(result);
+           result = NULL;
+           break;
+        }
+
+        switch (ret->type) {
+        case FLB_EXP_BOOL:
+            result->val.boolean = ret->val.boolean;
+            break;
+        case FLB_EXP_INT:
+            result->val.i64 = ret->val.i64;
+            break;
+        case FLB_EXP_FLOAT:
+                result->val.f64 = ret->val.f64;
+                break;
+            case FLB_EXP_STRING:
+                result->val.string = ret->val.string;
+        }
+        result->type = ret->type;
+        flb_free(ret);
+        break;
+    case FLB_LOGICAL_OP:
+        left = reduce_expression(expression->left, map);
```

Comment:
please validate return values since it can be NULL.

Reply:
Left and right sides of logical operations can be `NULL` and it will be interpreted differently depending on the operation. For instance, in the case of comparison, both sides have to contain a value, and in the case of `AND`, `OR`, and `NOT`, we interpret the `NULL` value as `false`. I missed the check for parenthesis, which submitted a fix.

Allowing a `NULL` value also makes the parser and the code simpler. For instance, we convert a condition on a single variable (i.e. `WHERE code`) to an operation (`WHERE code OR NULL`).